File: TechnicalAssessment/Question2/HighCardCompare.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Text;

namespace Question2
{
    public class HighCardCompare : IComparer<Card>
    {
        //:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
        #region Implement IComparer
        //:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
        //=====================================================================
        public int Compare([AllowNull] Card x, [AllowNull] Card y)
        {
            if (x.IsWildCard && y.IsWildCard)
                return 0;
            else if (x.IsWildCard)
                return -1;
            else if (y.IsWildCard)
                return 1;

            if (x.Number > y.Number)
                return -1;
            else if (x.Number < y.Number)
                return 1;
            else
            {
                if (!m_allowSite)
                    return 0;

                if (x.Suit < y.Suit)
                    return -1;
                
                if (x.Suit > y.Suit)
                    return 1;
            }

            return 0;                      
        }

        #endregion
        //:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
        #region Public
        //:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
        //=====================================================================
        public HighCardCompare(bool allowSite)
        {
            m_allowSite = allowSite;
        }

        #endregion
        //:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
        #region Fields
        //:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
        bool m_allowSite;

        #endregion
    }
}

File: TechnicalAssessment/F_Question2/HighCardCompare.fs
module HighCardCompare

open System.Collections.Generic

type HighCardComare (allowSite : bool) = 
    let m_allowSite = allowSite

    interface IComparer<Cards.Card> with
        member this.Compare(x: Cards.Card, y: Cards.Card): int =
        
            let compare (x : int, y : int, higherWin : bool) : int =
                match x > y with
                | true -> 1
                | false -> 
                    match x < y with
                    | true -> - 1
                    | false -> 0

            match x with
                | Cards.WildCard -> 
                    match y with
                        | Cards.WildCard -> 0
                        | _ -> 1
                | Cards.StandardCard card_x ->
                    match y with
                        | Cards.WildCard -> - 1
                        | Cards.StandardCard card_y ->
                            match compare(card_x.Number, card_y.Number, true) with
                                | result when result <> 0 -> result
                                | _ -> 
                                    match m_allowSite with
                                        | true -> compare(card_x.Suit, card_y.Suit, false)
                                        | false -> 0

type HighCardMultipleCompare (singleCompare : IComparer<Cards.Card>) =

    let m_singleCompare = singleCompare

    interface IComparer<Cards.Card list> with
        member this.Compare(x: Cards.Card list, y: Cards.Card list): int = 

            let rec compare (x: Cards.Card list, y: Cards.Card list, singleCompare : IComparer<Cards.Card>, initIndex : int) : int =
                match initIndex > (x.Length - 1) with
                | true -> 0
                | false ->
                    match m_singleCompare.Compare(x[initIndex], y[initIndex]) with
                    | 0 -> compare (x, y, singleCompare, initIndex + 1)
                    | x -> x

            match (x.Length > y.Length) with
            | true -> 1
            | false ->
                match (x.Length < y.Length) with
                | true -> -1
                | false -> compare(x, y, m_singleCompare, 0)



    





